src/main/resources/templates/db_scripts/hridLock.sql

```diff
@@ -0,0 +1,26 @@
+-- Reject any instance, holding and item change if the HRID changes.
+-- We cannot check this in Java because we don't load the old value
+-- for performance reasons.
+
+CREATE OR REPLACE FUNCTION lock_hrid() RETURNS TRIGGER AS $$
+BEGIN
+  IF NEW.jsonb->'hrid' IS DISTINCT FROM OLD.jsonb->'hrid' THEN
+      RAISE 'The hrid field cannot be changed: new=%, old=%. id=%',
+          NEW.jsonb->>'hrid', OLD.jsonb->>'hrid', NEW.id
+          USING ERRCODE = '23578', TABLE = TG_TABLE_NAME, SCHEMA = TG_TABLE_SCHEMA;
+  END IF;
+  RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+DROP TRIGGER IF EXISTS lock_hrid ON instance CASCADE;
+CREATE TRIGGER lock_hrid BEFORE UPDATE ON instance
+  FOR EACH ROW EXECUTE FUNCTION lock_hrid();
+
+DROP TRIGGER IF EXISTS lock_hrid ON holdings_record CASCADE;
+CREATE TRIGGER lock_hrid BEFORE UPDATE ON holdings_record
+  FOR EACH ROW EXECUTE FUNCTION lock_hrid();
+
+DROP TRIGGER IF EXISTS lock_hrid ON item CASCADE;
+CREATE TRIGGER lock_hrid BEFORE UPDATE ON item
+  FOR EACH ROW EXECUTE FUNCTION lock_hrid();
```

Comment:
I think this is an unfortunate design decision, it means that business logic is either shifted or duplicated in the database, which increases the maintenance and testing burden of logic like this.

Reply:
There is no duplication.

HridValidators.java is deleted and moved to hridLock.sql.

If HridValidators.java should be used these 6 APIs need to call it:

/instance-storage/instances/{instanceId}
/holdings-storage/holdings/{holdingsRecordId}
/item-storage/items/{itemId}
/instance-storage/batch/synchronous
/holdings-storage/batch/synchronous
/item-storage/batch/synchronous

If hridLock.sql is used only 3 triggers need to be enabled, they are automatically called.

If hridLock.sql is used the batch APIs can avoid a SELECT before running the upgrade. This gives a performance boost.

Please explain how this increases the maintenance and testing burden. Isn't it similar to foreign key and uniqueness checks that already exist for these tables?
Please explain why the maintenance and testing burden overrules the performance needs that libraries have requested for the 3 batch APIs.

Reply:
> If hridLock.sql is used the batch APIs can avoid a SELECT before running the upgrade. This gives a performance boost.

I understand the motivation.

> Please explain how this increases the maintenance and testing burden. 

This increases the maintenance burden because it increases the need for a developer to refer to both the java and database code to understand the behaviour of the system.

It increases the testing burden because this behaviour can now _only_ be tested using the database. Especially the business logic for denying the change of a HRID which relies on the coordination between the java and database code which makes testing in isolation harder.

This becomes more relevant when we consider the increasingly likely future of the inventory modules being merged.

> Isn't it similar to foreign key and uniqueness checks that already exist for these tables? 

I consider constraints in the database as a back stop and should be used sparingly. My line for this is crossed when business logic (denying the changing of a HRID) is implemented in a trigger. I think SQL and triggers are a poor way of expressing that kind of logic.

> Please explain why the maintenance and testing burden overrules the performance needs that libraries have requested for the 3 batch APIs.

I didn't make that assertion. I said I considered this change to be an unfortunate trade off.

I think this is another example of how FOLIO tends to make convenient small design decisions that end up not fitting in with the broader view. My understanding is that what folks want from these batch APIs is consistent and sustainable performance, which I think is different to outright speed (which I think is the argument made for this work).